core/packages/contracts/src/NativeTokens.sol

```diff
@@ -0,0 +1,149 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.19;
+
+import "openzeppelin/access/Ownable.sol";
+import "openzeppelin/access/AccessControl.sol";
+import "openzeppelin/token/ERC20/extensions/IERC20Metadata.sol";
+
+import "./TokenVault.sol";
+import "./SubstrateTypes.sol";
+import "./NativeTokensTypes.sol";
+import "./OutboundChannel.sol";
+
+/// @title Native Tokens
+/// @dev Manages locking, unlocking ERC20 tokens in the vault. Initializes ethereum native
+/// tokens on the substrate side via create.
+contract NativeTokens is AccessControl {
+    /// @dev Describes the type of message.
+    enum Action {
+        Unlock
+    }
+
+    /// @dev Message format.
+    struct Message {
+        /// @dev The action type.
+        Action action;
+        /// @dev The message payload.
+        bytes payload;
+    }
+
+    /// @dev Unlock payload format.
+    struct UnlockPayload {
+        /// @dev The ERC20 token to unlock.
+        address token;
+        /// @dev The destination address that will receive unlocked funds.
+        address recipient;
+        /// @dev The amount to unlock.
+        uint128 amount;
+    }
+
+    /// @dev Emitted once the funds are locked and a message is successfully queued.
+    event Locked(bytes recipient, address token, uint128 amount);
+
+    /// @dev Emitted once the funds are unlocked.
+    event Unlocked(address recipient, address token, uint128 amount);
+
+    /// @dev Emitted after enqueueing a a create token message to substrate.
+    event Created(address token);
+
+    /// @dev Set a new outbound channel.
+    event OutboundChannelUpdated(address newOutBoundChannel);
+
+    /* State */
+
+    bytes32 public constant ADMIN_ROLE = keccak256("ADMIN_ROLE");
+    bytes32 public constant SENDER_ROLE = keccak256("SENDER_ROLE");
+
+    bytes32 public immutable peerID;
+    bytes public peer;
+
+    TokenVault public immutable vault;
+    IOutboundChannel public outboundChannel;
+
+    uint256 public createTokenFee;
+
+    /* Errors */
+
+    error InvalidAmount();
+    error Unauthorized();
+    error NoFundsforCreateToken();
+
+    constructor(
+        TokenVault _vault,
+        IOutboundChannel _outboundChannel,
+        bytes memory _peer,
+        uint256 _createTokenFee
+    ) {
+        _grantRole(ADMIN_ROLE, msg.sender);
+        _setRoleAdmin(SENDER_ROLE, ADMIN_ROLE);
+        vault = _vault;
+        outboundChannel = _outboundChannel;
+        peer = _peer;
+        peerID = keccak256(_peer);
+        createTokenFee = _createTokenFee;
+    }
+
+    /// @dev Locks an amount of ERC20 Tokens in the vault and enqueues a mint message.
+    /// Requires the allowance to be set on the ERC20 token where the spender is the vault.
+    /// @param token The token to lock.
+    /// @param recipient The recipient on the substrate side.
+    /// @param amount The amount to lock.
+    function lock(address token, bytes calldata recipient, uint128 amount) external payable {
```

Comment:
what happens if i do a lock of a token before it's been created on the substrate side?

Reply:
You lose your money 😄 

There's no way to know whether a token was successfully created on the substrate side, unless we support message receipts. That's out of scope for our initial launch.

Also there are plenty of ways users could shoot themselves in the foot with this `lock` call (which by the way supports more routing parameters in a PR I'm working on). For example by submitting a bad recipient or routing parameter.

I'm expecting that frontends will help users supply the right inputs.

Reply:
Right - that sounds fine, as long as it doesn't mess with the future registration and locking of that token?